iree/hal/host/host_submission_queue.cc

```diff
@@ -176,113 +80,103 @@ Status HostSubmissionQueue::ProcessBatches(ExecuteFn execute_fn) {
         break;
       }
-      if (restart_iteration) break;
     }
   }
 
-  if (!permanent_error_.ok()) {
-    // If the sticky error got set while processing we need to abort all
-    // remaining submissions (simulating a device loss).
-    FailAllPending(permanent_error_);
-    return permanent_error_;
-  }
-
   return OkStatus();
 }
 
 Status HostSubmissionQueue::ProcessBatch(const PendingBatch& batch,
                                          const ExecuteFn& execute_fn) {
   IREE_TRACE_SCOPE0("HostSubmissionQueue::ProcessBatch");
 
-  // Complete the waits on all semaphores and reset them.
-  for (auto& semaphore_value : batch.wait_semaphores) {
-    if (semaphore_value.index() == 0) {
-      auto* binary_semaphore =
-          reinterpret_cast<HostBinarySemaphore*>(absl::get<0>(semaphore_value));
-      RETURN_IF_ERROR(binary_semaphore->EndWaiting());
-    } else {
-      // TODO(b/140141417): implement timeline semaphores.
-      return UnimplementedErrorBuilder(IREE_LOC) << "Timeline semaphores NYI";
-    }
-  }
+  // NOTE: the precondition is that the batch is ready to execute so we don't
+  // need to check the wait semaphores here.
 
   // Let the caller handle execution of the command buffers.
   RETURN_IF_ERROR(execute_fn(batch.command_buffers));
 
   // Signal all semaphores to allow them to unblock waiters.
-  for (auto& semaphore_value : batch.signal_semaphores) {
-    if (semaphore_value.index() == 0) {
-      auto* binary_semaphore =
-          reinterpret_cast<HostBinarySemaphore*>(absl::get<0>(semaphore_value));
-      RETURN_IF_ERROR(binary_semaphore->EndSignaling());
-    } else {
-      // TODO(b/140141417): implement timeline semaphores.
-      return UnimplementedErrorBuilder(IREE_LOC) << "Timeline semaphores NYI";
-    }
+  for (auto& signal_point : batch.signal_semaphores) {
+    auto* semaphore = reinterpret_cast<HostSemaphore*>(signal_point.semaphore);
+    RETURN_IF_ERROR(semaphore->Signal(signal_point.value));
   }
 
   return OkStatus();
 }
 
-Status HostSubmissionQueue::CompleteSubmission(Submission* submission,
-                                               Status status) {
+void HostSubmissionQueue::CompleteSubmission(Submission* submission,
+                                             Status status) {
   IREE_TRACE_SCOPE0("HostSubmissionQueue::CompleteSubmission");
 
-  // It's safe to drop any remaining batches - their semaphores will never be
-  // signaled but that's fine as we should be the only thing relying on them.
-  submission->pending_batches.clear();
-
-  // Signal the fence.
-  auto* fence = static_cast<HostFence*>(submission->fence.first);
   if (status.ok()) {
-    RETURN_IF_ERROR(fence->Signal(submission->fence.second));
+    if (!submission->pending_batches.empty()) {
+      // Completed with work remaining? Cause a failure.
+      status = FailedPreconditionErrorBuilder(IREE_LOC)
+               << "Submission ended prior to completion of all batches";
```

Comment:
Looks like this status isn't actually propagated.

How about
```
if (status.ok() && !submission->pending_batches.empty()) {
  // Completed with work remaining? Cause a failure.
  status = FailedPreconditionErrorBuilder(IREE_LOC)
             << "Submission ended prior to completion of all batches";
}
if (!status.ok()) {
  // Fail all pending ...
  ...
}

list_.take(submission).reset();
```

Reply:
good catch!